<a href="https://colab.research.google.com/github/vtseliousidiro/CB2330/blob/main/02_probability_foundations/exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Session 02 - Random Variables and Probability

[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Intertangler/CB2330-exercises/blob/main/02_probability_foundations/exercise.ipynb)

The second video ended on a page of formulas. Here we build them, on a real genome and on the
PCR test from the end of the video.

Sections 1, 2 and 3 have a line to fill in each, and section 4 has three. I've written the
sequences, the check cells and the figures already.

**Deliverable:** this notebook, run, committed to `cb2330-portfolio`.

## 0. Three phage genomes

The videos read the base probabilities of T4, T7 and M13 off a count. Below are the
sequences those counts came from.

M13 is here in full, 6407 bases. T4 and T7 are 169 kb and 40 kb, so the notebook carries
their first 11,000 bases. They came from NCBI: NC_000866.4, NC_001604.1 and NC_003287.2.

In [2]:
T4 = (
    "AATTTTCCTTATTAGGCCGCAAGGGCCTTCATAGTTTTAGCGATTTGGGAAACTTCATCATCACTTAAAG"
    "AGTTGCGATAACCGATGAAGTCGGAAACAATACGGAATTTCTTGGTAAACTCAGCAACCATTTTATCACT"
    "GTTTTTTGAAGCATTATTTGATAATACATCAAAAAGATTAGTTACTGTCCAAATGTCATGACCGATGGTA"
    "TCTTTTCCACCATTAAAATATACACCCTGTAATGAACTAACCATATTAGCGAGTCGTGTATATTCTTCAG"
    "AAACTTCATCTATACTGAAGTACTTCATCATAAAATCTAACTCAGGATACTTGATAATTTTATCAATATA"
    "TCGTTTAGCTGAACTTGAATAACCTACATACTTATCATAATCTACATCATCAAAAGCATCTACATATAAA"
    "TCACGCAAAGCTTCAAAAATACATTGGCACTGACCGAGTTCTTTTACCTTTTTCTGTAAAAGCGGACGAA"
    "TAACATAAAATTCATTAATGCCAATAAGATTAGCCATACGAATCAAAATATTCATAGATGGATGACAAAG"
    "AGATGTAGTACCATCCATAGAGAAAATATCAGAACGATGCATATACGCTACATAACCAGTAATTTCATCT"
    "GCTTCTGATGTGAGGCGTAAATAATTCCTCTTTTCCCAGCGCCCGTCTTTAATTTCAAACTTAAACGCTG"
    "TAGCAGCTTTAGGACGAGGAGCTTTACTTTTAACTACCTTTGGAATATAACTTTTTACTAAAGCTTCAAT"
    "TTCTGACAAATAATGAATGTTAACTTCATCACTTTCAAACATCGCCATAATATCAGGAAGCAAATCAATC"
    "TGCGATTCTACTTCTGGATTAATAAACAGAAGACGTTCGTTATGATGAATATTCAAAGTGTTATTAAATT"
    "CACTATCATCTAACGCACGTGCTAATCCACGGACAATATTAACACGATTTTTAATATTATCAATAACGAT"
    "ATTAATTTTTGTTGTATTAATACCAAACAGACGATAACTTGATGCAACGGCTGAAGTTTCATGACTTTGC"
    "TTAATGCGCTTCAGTCGAGGGTCAAGATTTACTTCATACACAACTCCCGCGTTGCATAACTTACTGTCAG"
    "GTTCAAACATGCTCTGCATCTTTTTATATGACAGATTTTTAGTCGTGAATTTGACTGAATTACTAATCAT"
    "ATAATCTCGAGCAGAATACCCCATCTTCATCAATTCACGATATGTGTGACGAGGAGATGTAGATTCTTTA"
    "AATCGTTTTACATCTTCATTAAATGCTTTCTCACTGAGTTCTTTAACTCGTTCAATAATATTTTTACGAG"
    "TGCGATCATCCAGTGAAAGAGCCTCGCGAGATGGAGCAATATCAAGTGAACCCATTGGAAACTTAATGTA"
    "ATTCACTTCATTGCGAATGCTTAGCCAGTTACGGTCTCTAATAACACCATCGATAGGATAAACAATACCA"
    "CCGTAGATAGCATATAATCCACCACGATCAGGCCAGTATCTTTCTGGATTTACACCGTAATAGTCATCAA"
    "AATCCGGAAAATAATCAATTTCGCGGTCAAGACCATTAATGATAGCCAAATCTTTGAACGGTCGCATGAT"
    "ATAAGAAACTTCATAAGCAAAGTTTCTAAAGTCTTTTTCTTCAACTGGAACTACGATTTCAATACCAGTT"
    "TTATCATCTGGACCCATTTCTTTTACGAATGTAGGTTTAATCTGTGGACCATCACCATCCATGTAAGCTA"
    "CATAACCACGAATTTCACCTTTATGATACGAAGTAATACTAAACGTATCAGTATAACTAAACGGAGATTT"
    "AGAACCTAAACCAAATCCGCCAATAAAGTCATTAGATTCAGCTTTAGATGAACTGAAGTATGAATTATAC"
    "AACCCAGGAGAATTATCATCACCTTGAATATCAAAATCACTCATACCCGGACCAAAATCTCGACAAACAA"
    "ATCGTGGGTCTAAACGTCCAGGAACTTGTATGATAAATTTTTCAGGATTTCCATTAAGTGCATGAGCATC"
    "AATCATGTTAGTAATCAATTCACGGACTACTGCGCGAATCTTGTTTGTATACAAATCAGATGACAGAATT"
    "TTAAATACTTTAGGAGATGCTGTGATGCTAAATGCTTTTGATTTAGAACCATTACCAAGAATTGTTTCTT"
    "TTTCAGTGGTGATAATCATAATTTCCTCATTAATTCATATTACGCTTAATAACTTCAGCAACTTCTAGTA"
    "GTTCATCTTTAGTTGCGGTGTCGGATTGAATTTTATCTCTAATATCTTTAAAGCGGGTTTTAAATTCTTC"
    "GGCTTCTCCCATATCGAAAAAGCGTTGAATGATTCTATATTCTCGATGAACTGCTTTATCAAAAAGTTCT"
    "AAATTTACTTTATATGATTTCATTTCAATATCCTCATTTGCCCAATTAATTATACCACATCCTTGTGGTA"
    "AAGTAAACTACTGGCTCATCCATTCTTTACGAAGGTCAGCATTATCTCCCATGAGCATTTCAAAAAGCTC"
    "TTTCCAGTTCTCAGGAAGTTTAACAACATCATATACTGGGTTTTGAATCATCTCACGATATTCAGATTTT"
    "TCCAAAGAGCCAAGTCCCTTAATATAACGGATGCTATGTTTAGGTAGAGCATCTTTGGCACTCTCATATT"
    "CAGCGACTGTATAAAACCATTCTTGTTTTTTACCGACCTGAGCGATGATTACAGGAGTTTTGACAAAGCG"
    "AATTCGTCCTTGCTCAAACAATTCTGGCCAATTACTAAAAAATCCGAGCAGAGAAGGATAAATAGAACCT"
    "AATCCAATAATCTCTTAATTATGAGGTATTTCTATAGATAGCCCGAAGGCTATCCATCGTGATCTGCGTC"
    "TGTCATAATAGCGACATTCGCATAGTTCATTGAAGAGGATTTAATAGAACGACGAACATTGTTCTGCAAT"
    "TTATATTTTTTCATATCAACGCTAGAAGAATCAATTTTTACAAATTTCATTATACACCTCATAGAACTTT"
    "TCATCAGGAATCCAACCGCGTTTAAATTCATTAAATGCTCGGCCGAATAATTTTGAATTCACAGTTATAT"
    "TATTAACTGATTTCCATTTAGCAACTCCCGTTCGTTTATAATGATCGGGGTCATATTTCGTGACGTACCA"
    "TTCATATAAATTTGGTATTAATTTTACAGCCTCTGGATTTGTCGCTGATTTATTATACCATGGTTTCGCT"
    "TTTTCAAGTTCAGCTGCTTTTGCAGTAGCAGCAACAGTATTTCCAACACCAACTAATTTTTCAGTTTTAA"
    "TGCGTGGATCATTAACATGAAGACGAAAAACTTCGCCCTTTTCATTTTTAAAGCATGTCATGCCTTTAAT"
    "TCCAGGAAGCTTAACACCTTTATTAACACCGCATCATTCCTTTGGGTTAAATGATCCTTTAATTAATAAG"
    "GCGCATTTACCCGATTTAACTACTTCTCATTCAACAACTTTATCTTTCATAACGTTTTTTGACCATTCAG"
    "ATACTGCTCTTTGATGGCTAAATTCTAGCAATTTCACTATAATTTGCACTAGAACGTAAACTATTTTTCA"
    "GTGTTTCAACATTCTATTCATCGCATATGCCATTTCACGATTACGATGAATTTTATATAGTAGAAAATAG"
    "TGCTAAAAAGTGTTCACGAAAAGTCATGTTTCACCAAATTTTCGTTATCATCAGAACCTCCCATTGATCT"
    "TGGAAGGATATGATGAATTTCACCCTTAAATTTAGAAACGCGGGTTTTACCCCGCACTATTAAGTCATTA"
    "TAGATTTTTTCGTAATTCACCTACTGTTATCCATTTACCATTAATCTGTACTTCATCATTTTCATTTACG"
    "ATAATTGTATCGCCATTTAGCTCGAAAGTAAACCACTCGCCATCTTCTTTTTCTTCAAACGCTTTTTCAC"
    "CGAGAACTAGACCAGTGATTGCGCAAATATCAAATAGTTCTTTGTTTTTAAGCATATCTGCATAAGACAT"
    "ACCCCAACTGTTGAGAACTTTACCACGCAATGGATAACCACCGTGAAGTTCTTTATCACGAACATCAATA"
    "AGATATCCGATAGCCGAATCACCCTCAGTCAAGAAAAGAGTAGTATCAGCATCTTTACCGCAAAGATTCG"
    "CTTTGATATGTTTATGAACCTTAGCTTTAGAAGCCTTTTTAGCTGCTTTAGTTTCTGCTGCTTTTTCTGC"
    "CGCCAATTTACGAGCCAAAGCAGCTTCAATAATCGGCATTAGAATTGCTTCATTATTTAGAATATCACGT"
    "GAAATCTTTTTAGCATCAAGTTGAATATGACTACGAATTTCGCCAAATGGAGAAGTCAAACGCTCTTTAG"
    "TTTGACGAATCAATCGCATGTTTTTCATATCACGAACAAACATAACGATAGTCAAACATTCTTTGACACG"
    "TGCTTTAGTCACATCAATTTTGAACTTACGTTTGATTTGTGGAATAAGGTCTTCACAAATATCATCCATA"
    "GCGCAGTCAATGTGATGGCCACCATTCTTAGTATGAATGTTATTGACGTATGTTAATTGACGAAAACCAT"
    "CCGGTGAACGACCAACCGCAATAGAACAATTTTCTTGCTCTTGAACAATAGCATGTTCATCATACTGCCG"
    "TGCATATTTCTTAAAATTGCCCTGAACCTTTTTACCATTAAAGGTAAATTGAATATCAGGATAAACTACA"
    "GCAAGTGTCTGGAGACGATCCAGTGTAATGTCAAGATAAACTTGGGACAGCTCATTAGTTTCAAATGACA"
    "TAAAATCAGGAATGAAAGTAACACGAGTTCCTTTCCATTTTCCAGGAATATCTTCCCATGATTTATTTTC"
    "CATGCCATTTGAACAACGAACTACAATATTATTTTGACCGTCGCCAGTTTCACCGACAAACATCACAGAA"
    "AAAATGTTTGTCAAACTAGAACCAACACCGTTCATACCGCCGGTGACGCGTTCTTTATCATCACCAAAGT"
    "TACCACCTGCTTTTGGAATAGTCCATGCGGCAACAGGACCAGGAATTTCTTCACCGGTAGGTGTTTTAAC"
    "CATCGCTTGTGGAATACCGCGACCGTTATCTTCAACTGTTACTTGATTGTTTTTAATAGTAACATTAATT"
    "TTATTCGCGAATTTAAACTTAGTACGAATACCTTCATCTACTGAGTTATCGATAATTTCATCAATAAGCT"
    "TAACAAGACCAGGTACATACTGAACACTTTCCCATTTACCAAACATAAAGCGCTCATGCGTTTCATTAGC"
    "AGAAGAGCCAATGTACATGCCACTACGCTTTTTGATATGTTCAATATCGCTCAGAATTTTAATTTCATTC"
    "TTAATCATCACTTATCCTCGTTTGGTTTCGGGAATATTATACTCCGGTAATCATAAAGCTAAAGGCCCGA"
    "AGGCCTTTTATTTAAAACGAATAGTTGAATCCTTAAAGAACAGCCCAGAACATACTGTTCCTTCTACTTT"
    "CTGCCCGGTAGGTCCAATAGCACGAAATCCAGTATGCTGGAAATCATTTTCAGAGCAACCGAACCAATTA"
    "TATCCAGTGATTTCAATATTAGTAAAACCACTTGAAGACAAAACTTTGGTTGCATTAATCAGCATCAGTA"
    "CTATTAATTAAAGACACTGCTAATACTAATGCTGCAATTGAACGACTAATATATTTCATAACTACCCTTT"
    "AAGCAAGTCGTAAAATCCATTATTCCCATGCTTAGGAAGCGGAAACTAACCGAACAGCCAGCCGATGACA"
    "ATCAGGACATACACCAGTATCTCTTCCAGAAATTTTCTTGATTTTTTCGTATTCTTTTGCACAGTCTTTG"
    "GATTGACATTTATAATCATAAAGCGGCATAATTATTCCTTAAAGTAAGCTTTCAACATCTGATATAAAGA"
    "CCACGCCTGATCATTATTTTCAATAGTAACTTTCATGACTGGGAATTCTGTGAAATCTTCTATTTGTTCT"
    "TGCTCTTTCTCTTCCTGCTCTTGCTCTTCAACCGCCTGATATGGATTTTCCACTTCATCAAAGAACCCAG"
    "CTTCGTTAGTAGAGAGCCAGATAAAGTTTTCGTCAAGGATATCACCGCCGGCACAACGTTTGAGTACACC"
    "TATGGATGTCATAATTTTAGTAGGACGCCCAAGATAATCAGCATCTAAAATTTTAAAAGGCTCCATACCT"
    "AAACGTCGTGCATAGATTCCGTTATCAGTATGGTCTTTAATAAAATTTTCTTGAGCTTGTTTATTTTTAA"
    "ATTGATACCATTTATTAACTTCAAATTTAATAGCCATTAATAAATTTCCTTCCAGTAAGTTGTGCCGTCT"
    "TCAGTAATTTCACGAAATACACCATAAATTGGCTGTTTATCACCGACTTTCTCATACACATAAACAGAAG"
    "TCAAGTGAGTAAACTTGCTAGTATGTTCCTTTTGAACTACTACCAAATTTGGATCAAATAATACATCTTC"
    "AAATTCATCATTAGTGCAATTCTGAACAATTTTACGTTTCATTACAATTTCCTCATTAATTGAACAGTGG"
    "AGCGATACGTTTCAGAAGAGTATCAACACCTTTAGCGAATTTTCCATTTTATTCTCCAAGTTGTTTTCTG"
    "TATCAGTAGTTGATATTGATATAGTACCATAATCAACTACTGATGTATATAGTTTTATGAAAAATTTAAA"
    "CTTTATGCATAGAGAGCATTGCTATAGTGTTTAATCCAACTTTCAGGAATGACTTTGTATGTTCCTAAAA"
    "ATACCACGTTGTACAACTTAACACCATCTTCTACCCATTGATCGGTAATGTATCCACACATAGCGCGAGT"
    "ATAAACAACCCTTCCATCATCTTTAATAAAGTTAAATTCACAAGGAGCAATGAACTTGATAGCCTGACCG"
    "AGTTTCCACTTAAAGTCTACACCTACATGCGAAGTATCAATCGTTTCAATTCCTTTAGCAGGAACAGCTT"
    "TTAAAAACGCAGACTCAAGAAATTTCGCACGAACATAGCCAAACTGGGGTTTAGACTTTCCATCTTTAGG"
    "AATGATACGCACTTTTACTTCAGAATCTTCATCTTTAACACCATGCTTAAGCTGAATGCTTACAACTTCG"
    "ACCAATTTTCCTGCTGCTTTAGAACGGGATTTATCAGATACACGAGCTAATTCACCAATATTAATAATCA"
    "TAGTTATCTCTCACTTGTTAAAAAGATTTTATACTCCACGGGACCATTATACTCTGGTCCCAAGAGTTTG"
    "TAAACTATTAATTCAAAATAGCTACCACTGCACTACGAGGAACTACGGAGTACTCTCCAGCATGAACTAC"
    "GTTCAGAAGTTCAACGCCATCTTCCAATCCATTGGTCAGTTACCCAACCACCAATCGGATTCGCAAATGG"
    "ACGACGAATGTAAACTGCCTTACACAACAAATCAGTCGGGTCTTCAGGTTTTTCAATTTCTGTCAATAGA"
    "TTAAAATCTTCTTCATAGATGATGAATGATGATACCCTTCCATAATAGTTTCTAATTTCCATGTACACTG"
    "TACCAATGAGAATTCCACTATTAACACTAATGACAGTAAAAGGATATCCGCCAGTTGTACCAAGAAGTTC"
    "CCAAAATTTGCTATCAGTCGTATTCGACATTGTCTGAAAATCAATTTCAGATGGTTTTTTCATTTGATAC"
    "GCAGTATTAATTTTAATCATAATTTTCTCTTTAGTTTAAGGTAATAAAGCCTTTTAGTTCGGCATAGGAT"
    "TTACGGAACATTACTTGATGCCCGCCAATGATAACTTGGTCATCTGGTACTTCGTATACAGCAAGATAAA"
    "ATCCTTTCGAAGCTAATTCTTCACGCTCTTCTCGTGTGAACCATTTCATCATATCATATTCGCTAGCAAA"
    "AGCAAAATGATAAAGAGCTACAAACCATCCGGGAATATGATATTCTACTCCAACATAATCTTTCTTGAAC"
    "TTAGTATTAATTACGATATTAGCATTTTTAACTAATAGTTTGTCTTCGTGCGGCACAGGAATTCTTTTAT"
    "TATCATTACTATGATGCATAAAATTAGGTCTGTCATAACCTACATGTAATAACCACTCTTCACTCCATGA"
    "ATCTATTATACTTCTATACGGCGTTATTTGAACACAAAGATCTCGGCGTATTGTTATAGCGTCTTCATAA"
    "TTAAGAATACTAAACGATGATTCAACACGATAAATTTTCATTTTATTATCCTCAGTAGCTATGGTGTTAT"
    "AGTACCACAACTAACCGAGGAAGTAAACAACTTTTTATCGTTTTGTTGGAAGAGATAGAGGATCGCATTC"
    "TTCCTCTGATGGAGCATCTTCAAGACCCATAGCATATCGCAAAGCATACTTCATCATCAGGATGTCTTTC"
    "GCACAGTCATGAATAGAATCATGTGCAACGAATCCATCTAAAGTTCCCTTTGGAAGAGGACACGTTGTCA"
    "TATCACGAACAAGCAGAAGTGCTTCAATTCTAGTACGAATATCACGCTGATTCCAAAATTTACAAGGTTC"
    "TAACTTAAATGTATCAAGCTCATTCTCGGAAACGCCGTTAAGACGTTGAATATCGCGAATAAGATCGACT"
    "AAAATTGGAAAATCAAACGACATTCCACGGCACCAGCCTTGAGATTTCCAAGGATCGATATTATGTGCAT"
    "TGATGTAATCATTAAATTTTGCAATACCGTCGATAGTGCTTACATCTTCATCGGATGGTGCAATATTTTT"
    "TCGAGCTTCAGGAGATTGATTCTTCCACCATTCGATAGTACTTTTAGTAAAAAGACGGTGTCCTTTTTGG"
    "CTTTTTAAATCAAATTTGATTTTAATGCCACGTGAAACTAATTCATCGAATGTTTCAACTACTTCTGGAT"
    "TAGGGTCAAAAGCAATTACAGCCAAATCAATAACCGCTGCTTTTTCACCACTTCCCATTGTTTCAAAATC"
    "TATAATAAAATCAAACATTAAATTTTCCTCGCTAAATCACGAATTTGACCTACAGTATAGTCTTGAATAT"
    "AAACTTTATTAATAGGCTCATCAATAAATTTTGCCATAGATTCAATATCTTTTTGTATTTCTTCAAGACT"
    "GTATACTATCTTTGAAGCTTTTTCGCGAATAGTGATATTTTCAGGACCCGGATTTTCTTGAATGACAACT"
    "TTAACATTTGTCATAAGAGATTTAAACTGGTACCAACTTAATTCAATCATTAATAATCGCCTCATAAAGA"
    "TAGCTAATTTCGCCTAAAACATAATCATTGATTGTAACAGTTTTAACTTCACCGCAAAAGAATTCTAACG"
    "CAATTAAATCTCGTTCAATTTCTTCTAATTGAAGCATCAACTTACTAGATTCAATTTTTACAGTTTCACG"
    "ATTTTTGCTATAAGCTATTTCATAAATTTCGCTTACTTTATCTTGAAGAAGATAAAACTGATCTTTAGTT"
    "ATTTCCACGAATAGCTTCCTCAAATTTAATCATACATAAAACACATCATAACGACCACGGGTGACACCAA"
    "CATAAAGAAGTTGTTGAGCTAATTCAACATCTGCATAATGAATACAAGGCGTATAAATGAAAGCACGGTC"
    "TACAGACATACCCTGCGCTTTATGGAATGTTGATGCAGGAAGTGCTTTCACTTTACTAAACTGTGATTTA"
    "GCATCCCAAAAATCACTCCACGGAGCTTTTCCGCCTTTGTTCCAATTTTTATAAGTTTCTGCTGTTTTAG"
    "CTAAAAATAGGTTAAACTTATACAATTCTTCGTCAGATGAAATTATTTTAATCTTTTCACGATAATATTC"
    "ATCATCGCCATAAGTTTCTACTGTTAAATCCCAATGACGAATTAGATATTCTCCAGGAACACCACGGGCT"
    "TTAACAAACGTTGATGTATACTCTGCTTCTATAATACGAACTAATTGTCCGTTATTAAAAATAATTTCTG"
    "ACACAGGCTTTCCATCAATTTTATATGTTTTAAATAATGGTTCCTGCATTACAATAATTTCACCGACAAT"
    "AAAATCTTTATCAGTTTCAAAAATCTTTTTACGAATAATGCTATTTAACTTGTCAACAGATTTATTCGTA"
    "AATGCCATTACGCGATTTTCAAACAAATCATCTAGTGATTTGACGATTGAAAAATAATTTACCATAAAAT"
    "CGCGTAAAGCGGTATCACCAGTAAATCCACGTACTCCATGCCCGTCAACAACTTTATCATAATTCCACTT"
    "ACCGTTGCGAACGTCAGTAGCTACATCAATAATAGGAGCATTACTGCGTTTAACTTCAGTGAGTTCACAC"
    "TGATAAAAATCTTTATGTGTAAAGAATGGACTGATATAAGCAGTATTTTCTCCTGGTTCAACAGGTCTGA"
    "TTTGCTTATTATCCCCTATTCCAATTATAGTACACCAAGGTGGAATAGTTGAAAGCAGAATTTTAAATAG"
    "CTTTCTATCATACATTGACACTTCGTCGCAGATTAATACTCTGCATTTGGCTAAATCAGGTACTTCTTTT"
    "TGTTCAAAAAGAACATTTTCTTCATATGTTACTGGGTTAATTTTAAGAATACTATGAATAGTACTCGCTT"
    "CTTTCCCTGATAGTTTTGAAAGAATCTTTTTAGCTGCATGTGTAGGAGCTGCTAAAATAATACCAGTTCC"
    "ACCCGTAGATATTAAAGCTTCAATGATGAACTTAGTAAGAGTAGTCTTACCGGTACCAGCAGGTCCATTA"
    "ATAGTTACATGATGTTTCTTTTCTTTAATAGCCTTCATAACAATGTTAAAGGCATTTTTCTGGCCTTCGG"
    "TCAAATCATCAAATGTCATCGTAAATTCCCTGCAATTGGTATACTAACAATACGCCCAGTATCTAAAATT"
    "CGCTGATATAATCTTTGCGTGTCTACGTCAGGCTTAACATGTTTAACTTCTATTTTATTAAACCAAAATT"
    "TACGTGGAGTCTCAACTAATCTTGGAATTCCCTTACCTAAAGCTAATCGATACTGCTCTTTAAGAGTGGT"
    "AAATACTTTATCAGCAATCTTCCATTCAAAAAATACAGCAGGACGATGTTCATCAAGCGGAACTGGCGCT"
    "GTAAATCCGT"
)

T7 = (
    "TCTCACAGTGTACGGACCTAAAGTTCCCCCATAGGGGGTACCTAAAGCCCAGCCAATCACCTAAAGTCAA"
    "CCTTCGGTTGACCTTGAGGGTTCCCTAAGGGTTGGGGATGACCCTTGGGTTTGTCTTTGGGTGTTACCTT"
    "GAGTGTCTCTCTGTGTCCCTATCTGTTACAGTCTCCTAAAGTATCCTCCTAAAGTCACCTCCTAACGTCC"
    "ATCCTAAAGCCAACACCTAAAGCCTACACCTAAAGACCCATCAAGTCAACGCCTATCTTAAAGTTTAAAC"
    "ATAAAGACCAGACCTAAAGACCAGACCTAAAGACACTACATAAAGACCAGACCTAAAGACGCCTTGTTGT"
    "TAGCCATAAAGTGATAACCTTTAATCATTGTCTTTATTAATACAACTCACTATAAGGAGAGACAACTTAA"
    "AGAGACTTAAAAGATTAATTTAAAATTTATCAAAAAGAGTATTGACTTAAAGTCTAACCTATAGGATACT"
    "TACAGCCATCGAGAGGGACACGGCGAATAGCCATCCCAATCGACACCGGGGTCAACCGGATAAGTAGACA"
    "GCCTGATAAGTCGCACGAAAAACAGGTATTGACAACATGAAGTAACATGCAGTAAGATACAAATCGCTAG"
    "GTAACACTAGCAGCGTCAACCGGGCGCACAGTGCCTTCTAGGTGACTTAAGCGCACCACGGCACATAAGG"
    "TGAAACAAAACGGTTGACAACATGAAGTAAACACGGTACGATGTACCACATGAAACGACAGTGAGTCACC"
    "ACACTGAAAGGTGATGCGGTCTAACGAAACCTGACCTAAGACGCTCTTTAACAATCTGGTAAATAGCTCT"
    "TGAGTGCATGACTAGCGGATAACTCAAGGGTATCGCAAGGTGCCCTTTATGATATTCACTAATAACTGCA"
    "CGAGGTAACACAAGATGGCTATGTCTAACATGACTTACAACAACGTTTTCGACCACGCTTACGAAATGCT"
    "GAAAGAAAACATCCGTTATGATGACATCCGTGACACTGATGACCTGCACGATGCTATTCACATGGCTGCC"
    "GATAATGCAGTTCCGCACTACTACGCTGACATCTTTAGCGTAATGGCAAGTGAGGGCATTGACCTTGAGT"
    "TCGAAGACTCTGGTCTGATGCCTGACACCAAGGACGTAATCCGCATCCTGCAAGCGCGTATCTATGAGCA"
    "ATTAACGATTGACCTCTGGGAAGACGCAGAAGACTTGCTCAATGAATACTTGGAGGAAGTCGAGGAGTAC"
    "GAGGAGGATGAAGAGTAATGTCTACTACCAACGTGCAATACGGTCTGACCGCTCAAACTGTACTTTTCTA"
    "TAGCGACATGGTGCGCTGTGGCTTTAACTGGTCACTCGCAATGGCACAGCTCAAAGAACTGTACGAAAAC"
    "AACAAGGCAATAGCTTTAGAATCTGCTGAGTGATAGACTCAAGGTCGCTCCTAGCGAGTGGCCTTTATGA"
    "TTATCACTTTACTTATGAGGGAGTAATGTATATGCTTACTATCGGTCTACTCACCGCTCTAGGTCTAGCT"
    "GTAGGTGCATCCTTTGGGAAGGCTTTAGGTGTAGCTGTAGGTTCCTACTTTACCGCTTGCATCATCATAG"
    "GAATCATCAAAGGGGCACTACGCAAATGATGAAGCACTACGTTATGCCAATCCACACGTCCAACGGGGCA"
    "ACCGTATGTACACCTGATGGGTTCGCAATGAAACAACGAATCGAACGCCTTAAGCGTGAACTCCGCATTA"
    "ACCGCAAGATTAACAAGATAGGTTCCGGCTATGACAGAACGCACTGATGGCTTAAAGAAAGGTTATATGC"
    "CCAATGGCACACTATACGCTGCAAATCGGCGAATAGTGAGAACTTGGCGAGAGAACAACCTCGAACGCCG"
    "CAAGGACAAGAGAGGGCGGCGTGGCATAGACGAAAGGAAAAGGTTAAAGCCAAGAAACTCGCCGCACTTG"
    "AACAGGCACTAGCCAACACACTGAACGCTATCTCATAACGAACATAAAGGACACAATGCAATGAACATTA"
    "CCGACATCATGAACGCTATCGACGCAATCAAAGCACTGCCAATCTGTGAACTTGACAAGCGTCAAGGTAT"
    "GCTTATCGACTTACTGGTCGAGATGGTCAACAGCGAGACGTGTGATGGCGAGCTAACCGAACTAAATCAG"
    "GCACTTGAGCATCAAGATTGGTGGACTACCTTGAAGTGTCTCACGGCTGACGCAGGGTTCAAGATGCTCG"
    "GTAATGGTCACTTCTCGGCTGCTTATAGTCACCCGCTGCTACCTAACAGAGTGATTAAGGTGGGCTTTAA"
    "GAAAGAGGATTCAGGCGCAGCCTATACCGCATTCTGCCGCATGTATCAGGGTCGTCCTGGTATCCCTAAC"
    "GTCTACGATGTACAGCGCCACGCTGGATGCTATACGGTGGTACTTGACGCACTTAAGGATTGCGAGCGTT"
    "TCAACAATGATGCCCATTATAAATACGCTGAGATTGCAAGCGACATCATTGATTGCAATTCGGATGAGCA"
    "TGATGAGTTAACTGGATGGGATGGTGAGTTTGTTGAAACTTGTAAACTAATCCGCAAGTTCTTTGAGGGC"
    "ATCGCCTCATTCGACATGCATAGCGGGAACATCATGTTCTCAAATGGAGACGTACCATACATCACCGACC"
    "CGGTATCATTCTCGCAGAAGAAAGACGGTGGCGCATTCAGCATCGACCCTGAGGAACTCATCAAGGAAGT"
    "CGAGGAAGTCGCACGACAGAAAGAAATTGACCGCGCTAAGGCCCGTAAAGAACGTCACGAGGGGCGCTTA"
    "GAGGCACGCAGATTCAAACGTCGCAACCGCAAGGCACGTAAAGCACACAAAGCTAAGCGCGAAAGAATGC"
    "TTGCTGCGTGGCGATGGGCTGAACGTCAAGAACGGCGTAACCATGAGGTAGCTGTAGATGTACTAGGAAG"
    "AACCAATAACGCTATGCTCTGGGTCAACATGTTCTCTGGGGACTTTAAGGCGCTTGAGGAACGAATCGCG"
    "CTGCACTGGCGTAATGCTGACCGGATGGCTATCGCTAATGGTCTTACGCTCAACATTGATAAGCAACTTG"
    "ACGCAATGTTAATGGGCTGATAGTCTTATCTTACAGGTCATCTGCGGGTGGCCTGAATAGGTACGATTTA"
    "CTAACTGGAAGAGGCACTAAATGAACACGATTAACATCGCTAAGAACGACTTCTCTGACATCGAACTGGC"
    "TGCTATCCCGTTCAACACTCTGGCTGACCATTACGGTGAGCGTTTAGCTCGCGAACAGTTGGCCCTTGAG"
    "CATGAGTCTTACGAGATGGGTGAAGCACGCTTCCGCAAGATGTTTGAGCGTCAACTTAAAGCTGGTGAGG"
    "TTGCGGATAACGCTGCCGCCAAGCCTCTCATCACTACCCTACTCCCTAAGATGATTGCACGCATCAACGA"
    "CTGGTTTGAGGAAGTGAAAGCTAAGCGCGGCAAGCGCCCGACAGCCTTCCAGTTCCTGCAAGAAATCAAG"
    "CCGGAAGCCGTAGCGTACATCACCATTAAGACCACTCTGGCTTGCCTAACCAGTGCTGACAATACAACCG"
    "TTCAGGCTGTAGCAAGCGCAATCGGTCGGGCCATTGAGGACGAGGCTCGCTTCGGTCGTATCCGTGACCT"
    "TGAAGCTAAGCACTTCAAGAAAAACGTTGAGGAACAACTCAACAAGCGCGTAGGGCACGTCTACAAGAAA"
    "GCATTTATGCAAGTTGTCGAGGCTGACATGCTCTCTAAGGGTCTACTCGGTGGCGAGGCGTGGTCTTCGT"
    "GGCATAAGGAAGACTCTATTCATGTAGGAGTACGCTGCATCGAGATGCTCATTGAGTCAACCGGAATGGT"
    "TAGCTTACACCGCCAAAATGCTGGCGTAGTAGGTCAAGACTCTGAGACTATCGAACTCGCACCTGAATAC"
    "GCTGAGGCTATCGCAACCCGTGCAGGTGCGCTGGCTGGCATCTCTCCGATGTTCCAACCTTGCGTAGTTC"
    "CTCCTAAGCCGTGGACTGGCATTACTGGTGGTGGCTATTGGGCTAACGGTCGTCGTCCTCTGGCGCTGGT"
    "GCGTACTCACAGTAAGAAAGCACTGATGCGCTACGAAGACGTTTACATGCCTGAGGTGTACAAAGCGATT"
    "AACATTGCGCAAAACACCGCATGGAAAATCAACAAGAAAGTCCTAGCGGTCGCCAACGTAATCACCAAGT"
    "GGAAGCATTGTCCGGTCGAGGACATCCCTGCGATTGAGCGTGAAGAACTCCCGATGAAACCGGAAGACAT"
    "CGACATGAATCCTGAGGCTCTCACCGCGTGGAAACGTGCTGCCGCTGCTGTGTACCGCAAGGACAAGGCT"
    "CGCAAGTCTCGCCGTATCAGCCTTGAGTTCATGCTTGAGCAAGCCAATAAGTTTGCTAACCATAAGGCCA"
    "TCTGGTTCCCTTACAACATGGACTGGCGCGGTCGTGTTTACGCTGTGTCAATGTTCAACCCGCAAGGTAA"
    "CGATATGACCAAAGGACTGCTTACGCTGGCGAAAGGTAAACCAATCGGTAAGGAAGGTTACTACTGGCTG"
    "AAAATCCACGGTGCAAACTGTGCGGGTGTCGATAAGGTTCCGTTCCCTGAGCGCATCAAGTTCATTGAGG"
    "AAAACCACGAGAACATCATGGCTTGCGCTAAGTCTCCACTGGAGAACACTTGGTGGGCTGAGCAAGATTC"
    "TCCGTTCTGCTTCCTTGCGTTCTGCTTTGAGTACGCTGGGGTACAGCACCACGGCCTGAGCTATAACTGC"
    "TCCCTTCCGCTGGCGTTTGACGGGTCTTGCTCTGGCATCCAGCACTTCTCCGCGATGCTCCGAGATGAGG"
    "TAGGTGGTCGCGCGGTTAACTTGCTTCCTAGTGAAACCGTTCAGGACATCTACGGGATTGTTGCTAAGAA"
    "AGTCAACGAGATTCTACAAGCAGACGCAATCAATGGGACCGATAACGAAGTAGTTACCGTGACCGATGAG"
    "AACACTGGTGAAATCTCTGAGAAAGTCAAGCTGGGCACTAAGGCACTGGCTGGTCAATGGCTGGCTTACG"
    "GTGTTACTCGCAGTGTGACTAAGCGTTCAGTCATGACGCTGGCTTACGGGTCCAAAGAGTTCGGCTTCCG"
    "TCAACAAGTGCTGGAAGATACCATTCAGCCAGCTATTGATTCCGGCAAGGGTCTGATGTTCACTCAGCCG"
    "AATCAGGCTGCTGGATACATGGCTAAGCTGATTTGGGAATCTGTGAGCGTGACGGTGGTAGCTGCGGTTG"
    "AAGCAATGAACTGGCTTAAGTCTGCTGCTAAGCTGCTGGCTGCTGAGGTCAAAGATAAGAAGACTGGAGA"
    "GATTCTTCGCAAGCGTTGCGCTGTGCATTGGGTAACTCCTGATGGTTTCCCTGTGTGGCAGGAATACAAG"
    "AAGCCTATTCAGACGCGCTTGAACCTGATGTTCCTCGGTCAGTTCCGCTTACAGCCTACCATTAACACCA"
    "ACAAAGATAGCGAGATTGATGCACACAAACAGGAGTCTGGTATCGCTCCTAACTTTGTACACAGCCAAGA"
    "CGGTAGCCACCTTCGTAAGACTGTAGTGTGGGCACACGAGAAGTACGGAATCGAATCTTTTGCACTGATT"
    "CACGACTCCTTCGGTACCATTCCGGCTGACGCTGCGAACCTGTTCAAAGCAGTGCGCGAAACTATGGTTG"
    "ACACATATGAGTCTTGTGATGTACTGGCTGATTTCTACGACCAGTTCGCTGACCAGTTGCACGAGTCTCA"
    "ATTGGACAAAATGCCAGCACTTCCGGCTAAAGGTAACTTGAACCTCCGTGACATCTTAGAGTCGGACTTC"
    "GCGTTCGCGTAACGCCAAATCAATACGACTCACTATAGAGGGACAAACTCAAGGTCATTCGCAAGAGTGG"
    "CCTTTATGATTGACCTTCTTCCGGTTAATACGACTCACTATAGGAGAACCTTAAGGTTTAACTTTAAGAC"
    "CCTTAAGTGTTAATTAGAGATTTAAATTAAAGAATTACTAAGAGAGGACTTTAAGTATGCGTAACTTCGA"
    "AAAGATGACCAAACGTTCTAACCGTAATGCTCGTGACTTCGAGGCAACCAAAGGTCGCAAGTTGAATAAG"
    "ACTAAGCGTGACCGCTCTCACAAGCGTAGCTGGGAGGGTCAGTAAGATGGGACGTTTATATAGTGGTAAT"
    "CTGGCAGCATTCAAGGCAGCAACAAACAAGCTGTTCCAGTTAGACTTAGCGGTCATTTATGATGACTGGT"
    "ATGATGCCTATACAAGAAAAGATTGCATACGGTTACGTATTGAGGACAGGAGTGGAAACCTGATTGATAC"
    "TAGCACCTTCTACCACCACGACGAGGACGTTCTGTTCAATATGTGTACTGATTGGTTGAACCATATGTAT"
    "GACCAGTTGAAGGACTGGAAGTAATACGACTCAGTATAGGGACAATGCTTAAGGTCGCTCTCTAGGAGTG"
    "GCCTTAGTCATTTAACCAATAGGAGATAAACATTATGATGAACATTAAGACTAACCCGTTTAAAGCCGTG"
    "TCTTTCGTAGAGTCTGCCATTAAGAAGGCTCTGGATAACGCTGGGTATCTTATCGCTGAAATCAAGTACG"
    "ATGGTGTACGCGGGAACATCTGCGTAGACAATACTGCTAACAGTTACTGGCTCTCTCGTGTATCTAAAAC"
    "GATTCCGGCACTGGAGCACTTAAACGGGTTTGATGTTCGCTGGAAGCGTCTACTGAACGATGACCGTTGC"
    "TTCTACAAAGATGGCTTTATGCTTGATGGGGAACTCATGGTCAAGGGCGTAGACTTTAACACAGGGTCCG"
    "GCCTACTGCGTACCAAATGGACTGACACGAAGAACCAAGAGTTCCATGAAGAGTTATTCGTTGAACCAAT"
    "CCGTAAGAAAGATAAAGTTCCCTTTAAGCTGCACACTGGACACCTTCACATAAAACTGTACGCTATCCTC"
    "CCGCTGCACATCGTGGAGTCTGGAGAAGACTGTGATGTCATGACGTTGCTCATGCAGGAACACGTTAAGA"
    "ACATGCTGCCTCTGCTACAGGAATACTTCCCTGAAATCGAATGGCAAGCGGCTGAATCTTACGAGGTCTA"
    "CGATATGGTAGAACTACAGCAACTGTACGAGCAGAAGCGAGCAGAAGGCCATGAGGGTCTCATTGTGAAA"
    "GACCCGATGTGTATCTATAAGCGCGGTAAGAAATCTGGCTGGTGGAAAATGAAACCTGAGAACGAAGCTG"
    "ACGGTATCATTCAGGGTCTGGTATGGGGTACAAAAGGTCTGGCTAATGAAGGTAAAGTGATTGGTTTTGA"
    "GGTGCTTCTTGAGAGTGGTCGTTTAGTTAACGCCACGAATATCTCTCGCGCCTTAATGGATGAGTTCACT"
    "GAGACAGTAAAAGAGGCCACCCTAAGTCAATGGGGATTCTTTAGCCCATACGGTATTGGCGACAACGATG"
    "CTTGTACTATTAACCCTTACGATGGCTGGGCGTGTCAAATTAGCTACATGGAGGAAACACCTGATGGCTC"
    "TTTGCGGCACCCATCGTTCGTAATGTTCCGTGGCACCGAGGACAACCCTCAAGAGAAAATGTAATCACAC"
    "TGGCTCACCTTCGGGTGGGCCTTTCTGCGTTTATAAGGAGACACTTTATGTTTAAGAAGGTTGGTAAATT"
    "CCTTGCGGCTTTGGCAGCTATCCTGACGCTTGCGTATATTCTTGCGGTATACCCTCAAGTAGCACTAGTA"
    "GTAGTTGGCGCTTGTTACTTAGCGGCAGTGTGTGCTTGCGTGTGGAGTATAGTTAACTGGTAATACGACT"
    "CACTAAAGGAGGTACACACCATGATGTACTTAATGCCATTACTCATCGTCATTGTAGGATGCCTTGCGCT"
    "CCACTGTAGCGATGATGATATGCCAGATGGTCACGCTTAATACGACTCACTAAAGGAGACACTATATGTT"
    "TCGACTTCATTACAACAAAAGCGTTAAGAATTTCACGGTTCGCCGTGCTGACCGTTCAATCGTATGTGCG"
    "AGCGAGCGCCGAGCTAAGATACCTCTTATTGGTAACACAGTTCCTTTGGCACCGAGCGTCCACATCATTA"
    "TCACCCGTGGTGACTTTGAGAAAGCAATAGACAAGAAACGTCCGGTTCTTAGTGTGGCAGTGACCCGCTT"
    "CCCGTTCGTCCGTCTGTTACTCAAACGAATCAAGGAGGTGTTCTGATGGGACTGTTAGATGGTGAAGCCT"
    "GGGAAAAAGAAAACCCGCCAGTACAAGCAACTGGGTGTATAGCTTGCTTAGAGAAAGATGACCGTTATCC"
    "ACACACCTGTAACAAAGGAGCTAACGATATGACCGAACGTGAACAAGAGATGATCATTAAGTTGATAGAC"
    "AATAATGAAGGTCGCCCAGATGATTTGAATGGCTGCGGTATTCTCTGCTCCAATGTCCCTTGCCACCTCT"
    "GCCCCGCAAATAACGATCAAAAGATAACCTTAGGTGAAATCCGAGCGATGGACCCACGTAAACCACATCT"
    "GAATAAACCTGAGGTAACTCCTACAGATGACCAGCCTTCCGCTGAGACAATCGAAGGTGTCACTAAGCCT"
    "TCCCACTACATGCTGTTTGACGACATTGAGGCTATCGAAGTGATTGCTCGTTCAATGACCGTTGAGCAGT"
    "TCAAGGGATACTGCTTCGGTAACATCTTAAAGTACAGACTACGTGCTGGTAAGAAGTCAGAGTTAGCGTA"
    "CTTAGAGAAAGACCTAGCGAAAGCAGACTTCTATAAAGAACTCTTTGAGAAACATAAGGATAAATGTTAT"
    "GCATAACTTCAAGTCAACCCCACCTGCCGACAGCCTATCTGATGACTTCACATCTTGCTCAGAGTGGTGC"
    "CGAAAGATGTGGGAAGAGACATTCGACGATGCGTACATCAAGCTGTATGAACTTTGGAAATCGAGAGGTC"
    "AATGACTATGTCAAACGTAAATACAGGTTCACTTAGTGTGGACAATAAGAAGTTTTGGGCTACCGTAGAG"
    "TCCTCGGAGCATTCCTTCGAGGTTCCAATCTACGCTGAGACCCTAGACGAAGCTCTGGAGTTAGCCGAAT"
    "GGCAATACGTTCCGGCTGGCTTTGAGGTTACTCGTGTGCGTCCTTGTGTAGCACCGAAGTAATACGACTC"
    "ACTATTAGGGAAGACTCCCTCTGAGAAACCAAACGAAACCTAAAGGAGATTAACATTATGGCTAAGAAGA"
    "TTTTCACCTCTGCGCTGGGTACCGCTGAACCTTACGCTTACATCGCCAAGCCGGACTACGGCAACGAAGA"
    "GCGTGGCTTTGGGAACCCTCGTGGTGTCTATAAAGTTGACCTGACTATTCCCAACAAAGACCCGCGCTGC"
    "CAGCGTATGGTCGATGAAATCGTGAAGTGTCACGAAGAGGCTTATGCTGCTGCCGTTGAGGAATACGAAG"
    "CTAATCCACCTGCTGTAGCTCGTGGTAAGAAACCGCTGAAACCGTATGAGGGTGACATGCCGTTCTTCGA"
    "TAACGGTGACGGTACGACTACCTTTAAGTTCAAATGCTACGCGTCTTTCCAAGACAAGAAGACCAAAGAG"
    "ACCAAGCACATCAATCTGGTTGTGGTTGACTCAAAAGGTAAGAAGATGGAAGACGTTCCGATTATCGGTG"
    "GTGGCTCTAAGCTGAAAGTTAAATATTCTCTGGTTCCATACAAGTGGAACACTGCTGTAGGTGCGAGCGT"
    "TAAGCTGCAACTGGAATCCGTGATGCTGGTCGAACTGGCTACCTTTGGTGGCGGTGAAGACGATTGGGCT"
    "GACGAAGTTGAAGAGAACGGCTATGTTGCCTCTGGTTCTGCCAAAGCGAGCAAACCACGCGACGAAGAAA"
    "GCTGGGACGAAGACGACGAAGAGTCCGAGGAAGCAGACGAAGACGGAGACTTCTAAGTGGAACTGCGGGA"
    "GAAAATCCTTGAGCGAATCAAGGTGACTTCCTCTGGGTGTTGGGAGTGGCAGGGCGCTACGAACAATAAA"
    "GGGTACGGGCAGGTGTGGTGCAGCAATACCGGAAAGGTTGTCTACTGTCATCGCGTAATGTCTAATGCTC"
    "CGAAAGGTTCTACCGTCCTGCACTCCTGTGATAATCCATTATGTTGTAACCCTGAACACCTATCCATAGG"
    "AACTCCAAAAGAGAACTCCACTGACATGGTAAATAAGGGTCGCTCACACAAGGGGTATAAACTTTCAGAC"
    "GAAGACGTAATGGCAATCATGGAGTCCAGCGAGTCCAATGTATCCTTAGCTCGCACCTATGGTGTCTCCC"
    "AACAGACTATTTGTGATATACGCAAAGGGAGGCGACATGGCAGGTTACGGCGCTAAAGGAATCCGAAAGG"
    "TTGGAGCGTTTCGCTCTGGCCTAGAGGACAAGGTTTCAAAGCAGTTGGAATCAAAAGGTATTAAATTCGA"
    "GTATGAAGAGTGGAAAGTGCCTTATGTAATTCCGGCGAGCAATCACACTTACACTCCAGACTTCTTACTT"
    "CCAAACGGTATATTCGTTGAGACAAAGGGTCTGTGGGAAAGCGATGATAGAAAGAAGCACTTATTAATTA"
    "GGGAGCAGCACCCCGAGCTAGACATCCGTATTGTCTTCTCAAGCTCACGTACTAAGTTATACAAAGGTTC"
    "TCCAACGTCTTATGGAGAGTTCTGCGAAAAGCATGGTATTAAGTTCGCTGATAAACTGATACCTGCTGAG"
    "TGGATAAAGGAACCCAAGAAGGAGGTCCCCTTTGATAGATTAAAAAGGAAAGGAGGAAAGAAATAATGGC"
    "TCGTGTACAGTTTAAACAACGTGAATCTACTGACGCAATCTTTGTTCACTGCTCGGCTACCAAGCCAAGT"
    "CAGAATGTTGGTGTCCGTGAGATTCGCCAGTGGCACAAAGAGCAGGGTTGGCTCGATGTGGGATACCACT"
    "TTATCATCAAGCGAGACGGTACTGTGGAGGCAGGACGAGATGAGATGGCTGTAGGCTCTCACGCTAAGGG"
    "TTACAACCACAACTCTATCGGCGTCTGCCTTGTTGGTGGTATCGACGATAAAGGTAAGTTCGACGCTAAC"
    "TTTACGCCAG"
)

M13 = (
    "AACGCTACTACTATTAGTAGAATTGATGCCACCTTTTCAGCTCGCGCCCCAAATGAAAATATAGCTAAAC"
    "AGGTTATTGACCATTTGCGAAATGTATCTAATGGTCAAACTAAATCTACTCGTTCGCAGAATTGGGAATC"
    "AACTGTTACATGGAATGAAACTTCCAGACACCGTACTTTAGTTGCATATTTAAAACATGTTGAGCTACAG"
    "CACCAGATTCAGCAATTAAGCTCTAAGCCATCCGCAAAAATGACCTCTTATCAAAAGGAGCAATTAAAGG"
    "TACTCTCTAATCCTGACCTGTTGGAGTTTGCTTCCGGTCTGGTTCGCTTTGAAGCTCGAATTAAAACGCG"
    "ATATTTGAAGTCTTTCGGGCTTCCTCTTAATCTTTTTGATGCAATCCGCTTTGCTTCTGACTATAATAGT"
    "CAGGGTAAAGACCTGATTTTTGATTTATGGTCATTCTCGTTTTCTGAACTGTTTAAAGCATTTGAGGGGG"
    "ATTCAATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCC"
    "CTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGT"
    "TATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTG"
    "GTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTTGTTCCGTTAGTTCGTTTTATTAA"
    "CGTAGATTTTTCTTCCCAACGTCCTGACTGGTATAATGAGCCAGTTCTTAAAATCGCATAAGGTAATTCA"
    "CAATGATTAAAGTTGAAATTAAACCATCTCAAGCCCAATTTACTACTCGTTCTGGTGTTTCTCGTCAGGG"
    "CAAGCCTTATTCACTGAATGAGCAGCTTTGTTACGTTGATTTGGGTAATGAATATCCGGTTCTTGTCAAG"
    "ATTACTCTTGATGAAGGTCAGCCAGCCTATGCGCCTGGTCTGTACACCGTTCATCTGTCCTCTTTCAAAG"
    "TTGGTCAGTTCGGTTCCCTTATGATTGACCGTCTGCGCCTCGTTCCGGCTAAGTAACATGGAGCAGGTCG"
    "CGGATTTCGACACAATTTATCAGGCGATGATACAAATCTCCGTTGTACTTTGTTTCGCGCTTGGTATAAT"
    "CGCTGGGGGTCAAAGATGAGTGTTTTAGTGTATTCTTTCGCCTCTTTCGTTTTAGGTTGGTGCCTTCGTA"
    "GTGGCATTACGTATTTTACCCGTTTAATGGAAACTTCCTCATGAAAAAGTCTTTAGTCCTCAAAGCCTCT"
    "GTAGCCGTTGCTACCCTCGTTCCGATGCTGTCTTTCGCTGCTGAGGGTGACGATCCCGCAAAAGCGGCCT"
    "TTAACTCCCTGCAAGCCTCAGCGACCGAATATATCGGTTATGCGTGGGCGATGGTTGTTGTCATTGTCGG"
    "CGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGCTGATAAACCGATACAATTAAA"
    "GGCTCCTTTTGGAGCCTTTTTTTTTGGAGATTTTCAACATGAAAAAATTATTATTCGCAATTCCTTTAGT"
    "TGTTCCTTTCTATTCTCACTCCGCTGAAACTGTTGAAAGTTGTTTAGCAAAACCCCATACAGAAAATTCA"
    "TTTACTAACGTCTGGAAAGACGACAAAACTTTAGATCGTTACGCTAACTATGAGGGTTGTCTGTGGAATG"
    "CTACAGGCGTTGTAGTTTGTACTGGTGACGAAACTCAGTGTTACGGTACATGGGTTCCTATTGGGCTTGC"
    "TATCCCTGAAAATGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTTCTGAGGGTGGCGGT"
    "ACTAAACCTCCTGAGTACGGTGATACACCTATTCCGGGCTATACTTATATCAACCCTCTCGACGGCACTT"
    "ATCCGCCTGGTACTGAGCAAAACCCCGCTAATCCTAATCCTTCTCTTGAGGAGTCTCAGCCTCTTAATAC"
    "TTTCATGTTTCAGAATAATAGGTTCCGAAATAGGCAGGGGGCATTAACTGTTTATACGGGCACTGTTACT"
    "CAAGGCACTGACCCCGTTAAAACTTATTACCAGTACACTCCTGTATCATCAAAAGCCATGTATGACGCTT"
    "ACTGGAACGGTAAATTCAGAGACTGCGCTTTCCATTCTGGCTTTAATGAGGATCCATTCGTTTGTGAATA"
    "TCAAGGCCAATCGTCTGACCTGCCTCAACCTCCTGTCAATGCTGGCGGCGGCTCTGGTGGTGGTTCTGGT"
    "GGCGGCTCTGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGCTCTGAGGGAGGCGGTTCCG"
    "GTGGTGGCTCTGGTTCCGGTGATTTTGATTATGAAAAGATGGCAAACGCTAATAAGGGGGCTATGACCGA"
    "AAATGCCGATGAAAACGCGCTACAGTCTGACGCTAAAGGCAAACTTGATTCTGTCGCTACTGATTACGGT"
    "GCTGCTATCGATGGTTTCATTGGTGACGTTTCCGGCCTTGCTAATGGTAATGGTGCTACTGGTGATTTTG"
    "CTGGCTCTAATTCCCAAATGGCTCAAGTCGGTGACGGTGATAATTCACCTTTAATGAATAATTTCCGTCA"
    "ATATTTACCTTCCCTCCCTCAATCGGTTGAATGTCGCCCTTTTGTCTTTAGCGCTGGTAAACCATATGAA"
    "TTTTCTATTGATTGTGACAAAATAAACTTATTCCGTGGTGTCTTTGCGTTTCTTTTATATGTTGCCACCT"
    "TTATGTATGTATTTTCTACGTTTGCTAACATACTGCGTAATAAGGAGTCTTAATCATGCCAGTTCTTTTG"
    "GGTATTCCGTTATTATTGCGTTTCCTCGGTTTCCTTCTGGTAACTTTGTTCGGCTATCTGCTTACTTTTC"
    "TTAAAAAGGGCTTCGGTAAGATAGCTATTGCTATTTCATTGTTTCTTGCTCTTATTATTGGGCTTAACTC"
    "AATTCTTGTGGGTTATCTCTCTGATATTAGCGCTCAATTACCCTCTGACTTTGTTCAGGGTGTTCAGTTA"
    "ATTCTCCCGTCTAATGCGCTTCCCTGTTTTTATGTTATTCTCTCTGTAAAGGCTGCTATTTTCATTTTTG"
    "ACGTTAAACAAAAAATCGTTTCTTATTTGGATTGGGATAAATAATATGGCTGTTTATTTTGTAACTGGCA"
    "AATTAGGCTCTGGAAAGACGCTCGTTAGCGTTGGTAAGATTCAGGATAAAATTGTAGCTGGGTGCAAAAT"
    "AGCAACTAATCTTGATTTAAGGCTTCAAAACCTCCCGCAAGTCGGGAGGTTCGCTAAAACGCCTCGCGTT"
    "CTTAGAATACCGGATAAGCCTTCTATATCTGATTTGCTTGCTATTGGGCGCGGTAATGATTCCTACGATG"
    "AAAATAAAAACGGCTTGCTTGTTCTCGATGAGTGCGGTACTTGGTTTAATACCCGTTCTTGGAATGATAA"
    "GGAAAGACAGCCGATTATTGATTGGTTTCTACATGCTCGTAAATTAGGATGGGATATTATTTTTCTTGTT"
    "CAGGACTTATCTATTGTTGATAAACAGGCGCGTTCTGCATTAGCTGAACATGTTGTTTATTGTCGTCGTC"
    "TGGACAGAATTACTTTACCTTTTGTCGGTACTTTATATTCTCTTATTACTGGCTCGAAAATGCCTCTGCC"
    "TAAATTACATGTTGGCGTTGTTAAATATGGCGATTCTCAATTAAGCCCTACTGTTGAGCGTTGGCTTTAT"
    "ACTGGTAAGAATTTGTATAACGCATATGATACTAAACAGGCTTTTTCTAGTAATTATGATTCCGGTGTTT"
    "ATTCTTATTTAACGCCTTATTTATCACACGGTCGGTATTTCAAACCATTAAATTTAGGTCAGAAGATGAA"
    "ATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTT"
    "ACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAGACCTATGATTTTGATA"
    "AATTCACTATTGACTCTTCTCAGCGTCTTAATCTAAGCTATCGCTATGTTTTCAAGGATTCTAAGGGAAA"
    "ATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCC"
    "ATTAAAAAAGGTAATTCAAATGAAATTGTTAAATGTAATTAATTTTGTTTTCTTGATGTTTGTTTCATCA"
    "TCTTCTTTTGCTCAGGTAATTGAAATGAATAATTCGCCTCTGCGCGATTTTGTAACTTGGTATTCAAAGC"
    "AATCAGGCGAATCCGTTATTGTTTCTCCCGATGTAAAAGGTACTGTTACTGTATATTCATCTGACGTTAA"
    "ACCTGAAAATCTACGCAATTTCTTTATTTCTGTTTTACGTGCTAATAATTTTGATATGGTTGGTTCAATT"
    "CCTTCCATAATTCAGAAGTATAATCCAAACAATCAGGATTATATTGATGAATTGCCATCATCTGATAATC"
    "AGGAATATGATGATAATTCCGCTCCTTCTGGTGGTTTCTTTGTTCCGCAAAATGATAATGTTACTCAAAC"
    "TTTTAAAATTAATAACGTTCGGGCAAAGGATTTAATACGAGTTGTCGAATTGTTTGTAAAGTCTAATACT"
    "TCTAAATCCTCAAATGTATTATCTATTGACGGCTCTAATCTATTAGTTGTTAGTGCACCTAAAGATATTT"
    "TAGATAACCTTCCTCAATTCCTTTCTACTGTTGATTTGCCAACTGACCAGATATTGATTGAGGGTTTGAT"
    "ATTTGAGGTTCAGCAAGGTGATGCTTTAGATTTTTCATTTGCTGCTGGCTCTCAGCGTGGCACTGTTGCA"
    "GGCGGTGTTAATACTGACCGCCTCACCTCTGTTTTATCTTCTGCTGGTGGTTCGTTCGGTATTTTTAATG"
    "GCGATGTTTTAGGGCTATCAGTTCGCGCATTAAAGACTAATAGCCATTCAAAAATATTGTCTGTGCCACG"
    "TATTCTTACGCTTTCAGGTCAGAAGGGTTCTATCTCTGTTGGCCAGAATGTCCCTTTTATTACTGGTCGT"
    "GTGACTGGTGAATCTGCCAATGTAAATAATCCATTTCAGACGATTGAGCGTCAAAATGTAGGTATTTCCA"
    "TGAGCGTTTTTCCTGTTGCAATGGCTGGCGGTAATATTGTTCTGGATATTACCAGCAAGGCCGATAGTTT"
    "GAGTTCTTCTACTCAGGCAAGTGATGTTATTACTAATCAAAGAAGTATTGCTACAACGGTTAATTTGCGT"
    "GATGGACAGACTCTTTTACTCGGTGGCCTCACTGATTATAAAAACACTTCTCAAGATTCTGGCGTACCGT"
    "TCCTGTCTAAAATCCCTTTAATCGGCCTCCTGTTTAGCTCCCGCTCTGATTCCAACGAGGAAAGCACGTT"
    "ATACGTGCTCGTCAAAGCAACCATAGTACGCGCCCTGTAGCGGCGCATTAAGCGCGGCGGGTGTGGTGGT"
    "TACGCGCAGCGTGACCGCTACACTTGCCAGCGCCCTAGCGCCCGCTCCTTTCGCTTTCTTCCCTTCCTTT"
    "CTCGCCACGTTCGCCGGCTTTCCCCGTCAAGCTCTAAATCGGGGGCTCCCTTTAGGGTTCCGATTTAGTG"
    "CTTTACGGCACCTCGACCCCAAAAAACTTGATTTGGGTGATGGTTCACGTAGTGGGCCATCGCCCTGATA"
    "GACGGTTTTTCGCCCTTTGACGTTGGAGTCCACGTTCTTTAATAGTGGACTCTTGTTCCAAACTGGAACA"
    "ACACTCAACCCTATCTCGGGCTATTCTTTTGATTTATAAGGGATTTTGCCGATTTCGGCCTATTGGTTAA"
    "AAAATGAGCTGATTTAACAAAAATTTAACGCGAATTTTAACAAAATATTAACGTTTACAATTTAAATATT"
    "TGCTTATACAATCTTCCTGTTTTTGGGGCTTTTCTGATTATCAACCGGGGTACATATGATTGACATGCTA"
    "GTTTTACGATTACCGTTCATCGATTCTCTTGTTTGCTCCAGACTCTCAGGCAATGACCTGATAGCCTTTG"
    "TAGACCTCTCAAAAATAGCTACCCTCTCCGGCATGAATTTATCAGCTAGAACGGTTGAATATCATATTGA"
    "TGGTGATTTGACTGTCTCCGGCCTTTCTCACCCTTTTGAATCTTTACCTACACATTACTCAGGCATTGCA"
    "TTTAAAATATATGAGGGTTCTAAAAATTTTTATCCTTGCGTTGAAATAAAGGCTTCTCCCGCAAAAGTAT"
    "TACAGGGTCATAATGTTTTTGGTACAACCGATTTAGCTTTATGCTCTGAGGCTTTATTGCTTAATTTTGC"
    "TAATTCTTTGCCTTGCCTGTATGATTTATTGGATGTT"
)

NAMES = ["T4", "T7", "M13"]
GENOMES = [T4, T7, M13]

for i in range(len(NAMES)):
    print(f"{NAMES[i]:4s} {len(GENOMES[i]):6d} bases   {GENOMES[i][:40]}...")

T4    11000 bases   AATTTTCCTTATTAGGCCGCAAGGGCCTTCATAGTTTTAG...
T7    11000 bases   TCTCACAGTGTACGGACCTAAAGTTCCCCCATAGGGGGTA...
M13    6407 bases   AACGCTACTACTATTAGTAGAATTGATGCCACCTTTTCAG...


A `str` behaves like a list of letters. `M13[0]` is the first base, `M13[0:10]` the first
ten, `len(M13)` how many there are, and `for letter in M13:` walks them in order.

---

# Part A. Build the four formulas

Sections 1 to 3 are functions that work apart from a removed line, marked
`⭐⭐⭐ Your code here ⭐⭐⭐`, and section 4 is Bayes' theorem written out three lines at a time.
A blank is a piece of a formula, never a piece of syntax. The check cell under each one runs it
on a hand-sized example and on the real numbers.

## 1. A probability, by counting

$$P\{A\} = \frac{\sum_{\omega \in \Omega} \mathbb{1}[\,\omega \in A\,]}
{\sum_{\omega \in \Omega} 1}$$

In a sequence the Laplace definition reads directly. $\Omega$ is the 6407 positions of M13,
$A$ is the event that a position holds a G, the sum on top counts the positions where that's
true, and the sum underneath is `len(seq)`.

The indicator $\mathbb{1}[\cdot]$ is an `if` and the sum around it is an accumulator. The
blank is the condition inside the `if`.

In [11]:
def p_base(seq, base):
    """P{A} for the event that a position holds `base`. Every position counts once."""
    n_hits = 0
    for letter in seq:
        # 1 line: the condition. True when `letter` is the base we're counting.
        if letter == base:
            n_hits = n_hits + 1
    return n_hits / len(seq)

In [12]:
assert p_base("ACGT", "A") == 0.25, p_base("ACGT", "A")
assert p_base("AAAA", "G") == 0.0, p_base("AAAA", "G")
assert abs(p_base(M13, "A") - 0.2460) < 5e-4, p_base(M13, "A")
assert abs(p_base(M13, "T") - 0.3467) < 5e-4, p_base(M13, "T")

total = 0.0
for b in "ACGT":
    total = total + p_base(M13, b)
assert abs(total - 1.0) < 1e-12, total

print("1 ok    P{A} =", round(p_base(M13, "A"), 4), " and the four sum to", round(total, 12))

1 ok    P{A} = 0.246  and the four sum to 1.0


The cell below prints the base probabilities per genome, with the GC content beside them.

In [13]:
for i in range(len(NAMES)):
    pA = p_base(GENOMES[i], "A")
    pC = p_base(GENOMES[i], "C")
    pG = p_base(GENOMES[i], "G")
    pT = p_base(GENOMES[i], "T")
    print(f"{NAMES[i]:4s}  A {pA:.3f}   C {pC:.3f}   G {pG:.3f}   T {pT:.3f}"
          f"    GC {pG + pC:.3f}")

T4    A 0.319   C 0.202   G 0.148   T 0.331    GC 0.349
T7    A 0.285   C 0.227   G 0.250   T 0.238    GC 0.477
M13   A 0.246   C 0.202   G 0.205   T 0.347    GC 0.407


Inside a single strand of a double-stranded genome, A and T come out at about the same
probability and so do C and G, which is Chargaff's second rule. T4 and T7 follow it.

M13 comes out at 0.246, 0.202, 0.205, 0.347 and breaks it, with A and T 10 points apart.
M13 is single-stranded, so it has no complementary strand to average against.

T4 and T7 come close to the drawn values rather than equal to them, because those came from
complete genomes and we're counting 11,000 bases.

## 2. When the outcomes are not equally likely

A fair die cuts the throw into six regions of the same size, and the Laplace definition gives
the regions equal weight. Load the die so a 6 comes up four times as often and counting still
gives $1/6$ where the answer is $4/9$, so the formula needs a weight:

$$P\{A\} = \frac{\sum_{\omega \in \Omega} w(\omega)\,\mathbb{1}[\,\omega \in A\,]}
{\sum_{\omega \in \Omega} w(\omega)}$$

The denominator adds up weights instead of counting outcomes, so the size of $\Omega$ is 9
rather than 6.

`p_weighted` takes a list of outcomes and a list of weights, the same length, so
`weights[i]` is the weight of `outcomes[i]`. Walking them together needs the position rather
than the item, which is what `range(len(outcomes))` gives us. `total` already accumulates the
denominator, and the blank is the numerator.

In [14]:
def p_weighted(outcomes, weights, event):
    """P{A} when outcome i carries weight w_i. `event` is the list of outcomes in A."""
    in_event = 0
    total = 0
    for i in range(len(outcomes)):
        total = total + weights[i]
        # 1 line. Add this outcome's weight to `in_event`. The `if` is given.
        if outcomes[i] in event:
            in_event = in_event + weights[i]
    return in_event / total

In [15]:
faces = [1, 2, 3, 4, 5, 6]
loaded = [1, 1, 1, 1, 1, 4]
fair = [1, 1, 1, 1, 1, 1]

assert abs(p_weighted(faces, loaded, [6]) - 4 / 9) < 1e-12, p_weighted(faces, loaded, [6])
assert abs(p_weighted(faces, loaded, [1]) - 1 / 9) < 1e-12, p_weighted(faces, loaded, [1])
assert abs(p_weighted(faces, fair, [6]) - 1 / 6) < 1e-12, "flat weights are the Laplace case"
assert p_weighted(faces, loaded, [1, 2, 3, 4, 5, 6]) == 1.0, "P{Omega} = 1"
assert p_weighted(faces, loaded, []) == 0.0, "P{} = 0"

print("2 ok    loaded die, P{6} =", round(p_weighted(faces, loaded, [6]), 4),
      " and P{even} =", round(p_weighted(faces, loaded, [2, 4, 6]), 4))

2 ok    loaded die, P{6} = 0.4444  and P{even} = 0.6667


The last checks are the first and second axioms: the impossible event comes out at 0 and the
certain event at 1, without either being built in.

A genome is loaded the same way. Its outcomes are A, C, G and T and its weights are the base
counts, so the cell below gets section 1's numbers back out of `p_weighted`.

In [16]:
bases = ["A", "C", "G", "T"]

counts = []
for b in bases:
    n = 0
    for letter in M13:
        if letter == b:
            n = n + 1
    counts.append(n)

print("M13 counts        ", counts, " summing to", sum(counts))
print("weighted P{G}     ", round(p_weighted(bases, counts, ["G"]), 4))
print("counted P{G}      ", round(p_base(M13, "G"), 4))
print("weighted P{G or C}", round(p_weighted(bases, counts, ["G", "C"]), 4))

M13 counts         [1576, 1296, 1314, 2221]  summing to 6407
weighted P{G}      0.2051
counted P{G}       0.2051
weighted P{G or C} 0.4074


## 3. Conditioning, and whether two events are independent

$$P\{A \mid B\} = \frac{P\{A \cap B\}}{P\{B\}}$$

Conditioning throws away the part of $\Omega$ where $B$ is false and renormalises what's
left. Put $B$ on a position and $A$ on the position after it: $B$ is "this base is a C",
$A$ is "the next base is a G", and $P\{A \mid B\}$ is the positions where a C is followed by
a G, over the positions holding a C.

The loop stops before the final position, because there is no base after it. `n_given`
already counts $B$, and the blank counts the positions where $A$ holds as well.

In [17]:
def p_next_given(seq, given, nxt):
    """P{A|B} for B = this base is `given` and A = the next base is `nxt`."""
    n_given = 0
    n_both = 0
    for i in range(len(seq) - 1):
        if seq[i] == given:
            n_given = n_given + 1
            # 1 line. Count this position into `n_both` as well. The `if` is given.
            if seq[i + 1] == nxt:
                n_both = n_both + 1
    return n_both / n_given

In [18]:
assert p_next_given("AAAT", "A", "A") == 2 / 3, p_next_given("AAAT", "A", "A")
assert p_next_given("AAAT", "A", "T") == 1 / 3, p_next_given("AAAT", "A", "T")
assert p_next_given("ACAC", "C", "A") == 1.0, p_next_given("ACAC", "C", "A")
assert abs(p_next_given(M13, "A", "A") - 0.3407) < 5e-4, p_next_given(M13, "A", "A")

row = 0.0
for b in "ACGT":
    row = row + p_next_given(M13, "A", b)
assert abs(row - 1.0) < 1e-12, row

print("3 ok    P{A|A} =", round(p_next_given(M13, "A", "A"), 4),
      " and the row sums to", round(row, 12))

3 ok    P{A|A} = 0.3407  and the row sums to 1.0


The row sums to 1: once we've conditioned on a C, the bases that can follow are again a
complete sample space. The cell below prints the table with section 1's unconditional
probabilities underneath.

In [19]:
print("            next base")
print("prev     A       C       G       T")
for prev in "ACGT":
    row = ""
    for nxt in "ACGT":
        row = row + f"  {p_next_given(M13, prev, nxt):.3f}"
    print(f" {prev}   {row}")

print()
marginal = ""
for b in "ACGT":
    marginal = marginal + f"  {p_base(M13, b):.3f}"
print(f" P{{x}}{marginal}      <- section 1, conditioning on nothing")

            next base
prev     A       C       G       T
 A     0.341  0.170  0.157  0.332
 C     0.208  0.204  0.205  0.383
 G     0.219  0.240  0.248  0.292
 T     0.217  0.202  0.214  0.368

 P{x}  0.246  0.202  0.205  0.347      <- section 1, conditioning on nothing


An A follows an A 34.1% of the time and a G follows an A 15.7%, where the unconditional
numbers are 24.6% and 20.5%. Independence means $P\{A \mid B\} = P\{A\}$, and M13 fails it
row by row: adjacent bases are correlated. Part B has a route that puts the correlation to
work and one that ignores it, and the two disagree.

---

## 4. A positive test, and what it is worth

Bayes' theorem with $H$ for the hypothesis and $E$ for the evidence:

$$P\{H \mid E\} = \frac{P\{E \mid H\}\,P\{H\}}
{\sum_{H'} P\{E \mid H'\}\,P\{H'\}}$$

The names, in the order this section builds them:

- the **prior** $P\{H\}$, how probable a hypothesis was before the experiment
- the **likelihood** $P\{E \mid H\}$, how probable this evidence is if that hypothesis is true
- the **joint** $P\{E \cap H\}$, the top of the fraction: this hypothesis holds *and* the
  evidence came out the way it did
- the **marginal** $P\{E\}$, the bottom: how probable the evidence is whatever produced it
- the **posterior** $P\{H \mid E\}$, what we believe once the evidence is in

No table lists the marginal, so we marginalise: split $E$ across a set of hypotheses that
covers everything, and add up the joints.

The video ran a PCR test through that formula. The hypotheses are infected and healthy, and
the evidence is a positive result. We do it twice. Section 4a writes the formula out as named
variables, which is the version that can be read straight off the page, and 4b is the same
arithmetic rewritten so that a third hypothesis is free.

### The numbers a test comes with

A diagnostic test comes advertised on sensitivity and specificity. **Sensitivity** is
$P\{+ \mid \text{infected}\}$, how often it catches an infection, 95% here. **Specificity**
is $P\{- \mid \text{healthy}\}$, how often it clears a healthy person, 99% here, so a healthy
person tests positive 1% of the time. Sensitivity and 1 minus specificity are the likelihoods of
a positive result under the hypotheses.

The prior is the **prevalence**, 1 person in 1000. The hypotheses have to cover everything
between them, so the other prior is whatever is left over.

In [20]:
SENSITIVITY = 0.95
SPECIFICITY = 0.99

# P{E|H}, the likelihood of a positive result under each hypothesis.
p_pos_given_infected = SENSITIVITY        # the test catches an infection
p_pos_given_healthy = 1 - SPECIFICITY     # a healthy person tests positive anyway

# P{H}, the prior: the prevalence, and what is left over.
p_infected = 0.001
p_healthy = 1 - p_infected

print(f"P{{+ | infected}} {p_pos_given_infected:.2f}      P{{+ | healthy}} {p_pos_given_healthy:.2f}")
print(f"P{{infected}}     {p_infected:.3f}     P{{healthy}}     {p_healthy:.3f}")

P{+ | infected} 0.95      P{+ | healthy} 0.01
P{infected}     0.001     P{healthy}     0.999


### 4a. One variable per symbol

Every piece of the formula gets a name of its own, so that a line of code can sit under it:

| In the formula | Called | In the code |
|---|---|---|
| $P\{H\}$ | the prior | `p_infected`, `p_healthy` |
| $P\{E \mid H\}$ | the likelihood | `p_pos_given_infected`, `p_pos_given_healthy` |
| $P\{E \mid H\}\,P\{H\}$ | the joint | `joint_infected`, `joint_healthy` |
| $\sum_{H'} P\{E \mid H'\}\,P\{H'\}$ | the marginal | `p_positive` |
| $P\{H \mid E\}$ | the posterior | `p_infected_given_pos` |

The prior and the likelihood are set in the cell above. The blanks below are the joint, the
marginal and the posterior, in that order: a multiplication, an addition, a division.

I have written `joint_healthy` out already as the model. The line under it is that same
multiplication with the other hypothesis in it.

In [23]:
# Stage 1, the numerator. P{E and H} = P{E|H} * P{H}, once per hypothesis.
joint_healthy = p_pos_given_healthy * p_healthy

# 1 line. The same multiplication for the infected hypothesis.
joint_infected = p_pos_given_infected * p_infected

# 1 line. Stage 2, the denominator. P{E} is the two joints added up.
p_positive = joint_healthy + joint_infected

# 1 line. Stage 3. P{H|E} is the infected joint over P{E}.
p_infected_given_pos = joint_infected / p_positive

In [24]:
assert abs(joint_infected - 0.00095) < 1e-9, joint_infected
assert abs(joint_healthy - 0.00999) < 1e-9, joint_healthy
assert abs(p_positive - 0.01094) < 1e-9, p_positive
assert abs(p_infected_given_pos - 95 / 1094) < 5e-4, p_infected_given_pos

print(f"P{{+ and infected}}   {joint_infected:.5f}")
print(f"P{{+ and healthy}}    {joint_healthy:.5f}")
print(f"P{{+}}                {p_positive:.5f}    <- the marginal, drawn as 0.01094")
print(f"P{{infected | +}}     {p_infected_given_pos:.4f}")
print("4a ok")

P{+ and infected}   0.00095
P{+ and healthy}    0.00999
P{+}                0.01094    <- the marginal, drawn as 0.01094
P{infected | +}     0.0868
4a ok


A test that catches 95% of infections comes back positive, and the person it came from is
8.7% likely to be infected.

The video got there by counting people instead. Out of 100,000 tested, 100 are infected and 95
of them test positive, while the 99,900 healthy ones produce 999 false positives, so 95
positives out of 1094. Every number on that page is a count of people, and `p_positive` is
where the count went: 0.01094 is 1094 positives per 100,000 tests. The false positives
outnumber the true ones by a factor of 10, and that ratio is the answer.

The healthy joint gets built too, and the denominator is where it appears.

### 4b. The same three lines, for any number of hypotheses

A pair of hypotheses fits in named variables. A longer list does not: `joint_infected` and
`joint_healthy` would turn into `joint_h1`, `joint_h2`, `joint_h3`, and `p_positive` would be a
hand-written sum that grows a term every time a hypothesis is added.

A `dict` holds values under names instead of under positions, which fixes that.
`priors["infected"]` reads the entry for that hypothesis, `priors["infected"] = 0.001` writes
it, and `for name in priors:` walks the names, however many of them there are. So the pairs of
variables from 4a become dictionaries with an entry per hypothesis:

| 4a | 4b |
|---|---|
| `p_pos_given_infected`, `p_pos_given_healthy` | `likelihoods[name]` |
| `p_infected`, `p_healthy` | `priors[name]` |
| `joint_infected`, `joint_healthy` | `joint[name]` |
| `p_positive` | `evidence` |
| `p_infected_given_pos` | `result[name]` |

`posterior` below is written out rather than blanked, because it is the lines from 4a with the
names swapped and a `for` loop wrapped around them: the multiplication, then the addition, then
the division, in that order. The 4a cell is worth rereading beside it.

A name that is not in the dictionary raises a `KeyError` rather than quietly returning
something wrong, and the cause is usually a typo.

In [25]:
# The same four numbers, keyed by hypothesis instead of spelled out one variable at a time.
LIKELIHOOD = {"infected": p_pos_given_infected, "healthy": p_pos_given_healthy}
PREVALENCE = {"infected": p_infected, "healthy": p_healthy}


def posterior(likelihoods, priors):
    """P{H|E} for every hypothesis in `likelihoods`, as a dict summing to 1."""
    joint = {}
    for name in likelihoods:
        joint[name] = likelihoods[name] * priors[name]      # stage 1, the numerator

    evidence = 0.0
    for name in joint:
        evidence = evidence + joint[name]                   # stage 2, the marginal

    result = {}
    for name in joint:
        result[name] = joint[name] / evidence               # stage 3, the division
    return result


for name in LIKELIHOOD:
    print(f"{name:9s}   likelihood {LIKELIHOOD[name]:.2f}   prior {PREVALENCE[name]:.3f}")

infected    likelihood 0.95   prior 0.001
healthy     likelihood 0.01   prior 0.999


In [26]:
p = posterior(LIKELIHOOD, PREVALENCE)

total = 0.0
for name in p:
    total = total + p[name]
assert abs(total - 1.0) < 1e-12, total
assert abs(p["infected"] - p_infected_given_pos) < 1e-12, (p, p_infected_given_pos)

dead = posterior(LIKELIHOOD, {"infected": 0.0, "healthy": 1.0})
assert dead["infected"] == 0.0, "a prior of zero has to stay at zero"

print("4b ok   P{infected | +} =", round(p["infected"], 4),
      " and it matches 4a to the last digit")
print("        the posteriors sum to", round(total, 12))

4b ok   P{infected | +} = 0.0868  and it matches 4a to the last digit
        the posteriors sum to 1.0


The function reproduces 4a to the last digit, which is why it is worth running: no arithmetic
was added, and the lines were rewritten so that the number of hypotheses stopped mattering.

The posteriors sum to 1 because the joints they came from add up to the number sitting
underneath them. And a prior of zero stays at zero whatever the evidence says, because nothing
recovers from a multiplication by zero.

### 4c. Three hypotheses, and what they stand for

The video finishes on a second example: hypotheses drawn as coloured bubbles, called $H_1$,
$H_2$ and $H_3$, and an experiment whose result is $a$, $b$ or $c$. They are left abstract on
the page because the arithmetic does not care what they are. Here is a clinic that runs on the
same numbers.

A patient comes in with a fever. The explanations on the table are an ordinary cold virus,
influenza and a bacterial throat infection, and suppose that among this week's feverish patients
they run at 50%, 30% and 20%. Those fractions are the priors, and they are the video's 50, 30
and 20.

The evidence is the answer to a question: did the fever come on within a few hours, or over a
couple of days? Abrupt onset is characteristic of flu, and the likelihoods say so. Suppose it
happens in half of the flu cases, in 10% of the colds and in 5% of the bacterial ones, which
are the video's likelihoods for result $a$.

So the evidence is 5 times more probable under flu than under a cold, and the prior runs the
other way and favours the cold by 1.7. The posterior is where they meet. `posterior` itself
does not change: the same function, with a name per cause.

In [27]:
# P{H}: the mix of causes among this week's feverish patients.
CAUSES = {"influenza": 0.30, "cold": 0.50, "bacterial": 0.20}

# P{E|H}: how often the fever comes on within a few hours, under each cause.
ABRUPT = {"influenza": 0.50, "cold": 0.10, "bacterial": 0.05}

after = posterior(ABRUPT, CAUSES)

print("cause         prior    joint    posterior")
for name in CAUSES:
    print(f"{name:11s}   {CAUSES[name]:.2f}    "
          f"{ABRUPT[name] * CAUSES[name]:.3f}      {after[name]:.3f}")

cause         prior    joint    posterior
influenza     0.30    0.150      0.714
cold          0.50    0.050      0.238
bacterial     0.20    0.010      0.048


In [28]:
assert abs(after["influenza"] - 0.7143) < 1e-3, after
assert abs(after["cold"] - 0.2381) < 1e-3, after
assert abs(after["bacterial"] - 0.0476) < 1e-3, after

total = 0.0
joints = 0.0
for name in after:
    total = total + after[name]
    joints = joints + ABRUPT[name] * CAUSES[name]
assert abs(total - 1.0) < 1e-12, total

print("4c ok   the three joints add to", round(joints, 3),
      "and the posteriors to", round(total, 12))

4c ok   the three joints add to 0.21 and the posteriors to 1.0


The joints come out at 0.15, 0.05 and 0.01, which are the boxes drawn on the page, and they
add to 0.21 rather than to 1. Dividing by 0.21 is the renormalisation, and the same division
turned 0.00095 into 0.087 for the PCR test.

The cold went into the experiment as the leading explanation at 0.50 and comes out second at
0.24. Flu went in at 0.30 and comes out at 0.71. The question reordered them, and a reordering
on new evidence is a Bayesian update: the priors are what we believed, the posteriors are what
we believe now, and the experiment is what happened in between. The video calls that the logic
of the experimental method, and it runs on the arithmetic of section 4a.

---

## 5. What the prior is doing

The test never changes. The cell below runs the same two likelihoods against a prevalence
climbing from 1 in 1000 to 1 in 2.

In [29]:
print("prevalence    P{infected | positive}")
for prevalence in [0.001, 0.01, 0.05, 0.10, 0.50]:
    priors = {"infected": prevalence, "healthy": 1 - prevalence}
    p = posterior(LIKELIHOOD, priors)
    print(f"  {prevalence:8.3f}    {p['infected']:.3f}")

prevalence    P{infected | positive}
     0.001    0.087
     0.010    0.490
     0.050    0.833
     0.100    0.913
     0.500    0.990


One positive result is worth 0.087 in a screening programme and 0.990 in a clinic where half
the people who walk in are infected. Between them the answer crosses 0.5 at a prevalence near
1 in 96, and that is where a positive result becomes more likely right than wrong.

So a sensitivity and a specificity don't tell us what a positive result means. The prevalence
does, and a posterior reported without the prior it came from leaves a reader unable to tell
which of these lines they're being shown. Missing that is common enough to have a name, and
the video draws it in red: base rate neglect.

---

# Part B (optional). Push it until it breaks

Part B is optional and it isn't assessed, so it sits outside the session for anyone who wants
more practice with the four formulas. One of the routes below is enough on its own.

**1. Classify a sequencing read.** `phages.fasta` holds T4, T7 and M13, and section 1 already
measures the base frequencies that separate them. Treat each phage as a hypothesis, build its
four base probabilities with `p_base`, and take the likelihood of a 40-base read under a phage
to be $\prod_{i} p_H(b_i)$, the product of the probabilities that phage gives to the bases we
saw. That product is an accumulator like section 1's, starting at 1.0 and multiplying rather
than adding. Feeding the three likelihoods to `posterior` classifies the read. A classifier
that multiplies the bases as though they were drawn separately is called naive Bayes, and
section 3 measured that they aren't, so its likelihoods come out more extreme than they should
be. It's used in production anyway, in spam filtering and in read binning, because the ranking
of the hypotheses often survives an assumption the individual numbers don't.

**2. Then run a prior against it.** Take a read from M13 and classify it twice, once with the
three phages equally likely and once with a prep that is 80% T7, 15% T4 and 5% M13. Bases 1431
to 1471 of M13 go to M13 at 0.79 under the flat prior and to T7 at 0.80 under the skewed one,
on identical likelihoods: the read favours M13 by a factor of 4 and the prior favours T7 by a
factor of 16. Reads long enough to settle it break the arithmetic instead, and a 400-base
product falls under $10^{-308}$ and comes back as 0.0 for all three, which is a real limit of
this implementation and has a standard fix with a name.

**3. Condition on the previous base.** Replace $\prod_i p_H(b_i)$ with
$p_H(b_1)\prod_{i=2}^{L} P_H\{b_i \mid b_{i-1}\}$, which `p_next_given` already computes and
which needs 16 numbers per phage instead of 4. Classifying the same reads both ways shows
whether the extra 12 numbers bought anything.

**4. Move the test instead of the population.** Section 5 moves the prevalence and holds the
test fixed. Hold the prevalence at 1 in 1000 instead and find what specificity a test needs
before a positive result is more likely right than wrong, then do the same for sensitivity.
One of the two barely matters.

**5. A conditional probability from our own field.** Find a dataset with a pair of categorical
columns, or a table of counts cross-classified by two of them, and compute $P\{A \mid B\}$,
$P\{B \mid A\}$ and $P\{A\}$ from the counts. $P\{A \mid B\} = P\{A\}$ is then the
independence check from section 3.